### ISSUE TITLE
Non-negligible overhead of `OpOverloadPacket` dispatch w/o overload

### ISSUE THREAD

**Issue by EikanWang** [pytorch/pytorch#153626]
### 🐛 Describe the bug

Dispatching an `OpOverloadPacket` to the correct overload may incur non-negligible overhead when users invoke it without explicitly specifying the overload name.

For example, suppose a Torch extension defines an operation `abc` under the `foo` namespace with two overloads:

- `abc.scalarint(Tensor self, int other) -> Tensor`
- `abc.Tensor(Tensor self, Tensor other) -> Tensor`

If a Torch script program invokes `abc` without specifying the overload explicitly

```python
import torch
a = torch.randn(10, device="cuda")
b = torch.randn(10, device="cuda")
c = torch.ops.foo.abc(a, b)
```

The dispatch mechanism needs to enumerate all available overloads and determine the appropriate one based on the schema:
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.cpp#L788-L796

Let’s examine how the dispatcher checks whether `abc.scalarint` is a valid candidate.

First, it retrieves the schema of `abc.scalarint`. And then, each argument is evaluated against the schema using `argumentToIValue`, which attempts to convert the Python argument to the expected schema type.
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.h#L1028-L1044

If the argument is incompatible, a C++ exception is raised.
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.h#L808-L834

In the example above, the second argument is a Tensor, but `abc.scalarint` expects an `int`. This mismatch triggers a `py::cast_error` exception.
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.h#L816-L823

To provide an informative error message, `py::repr(object)` is called, **which leads to the actual overhead: invoking `__repr__` on a Tensor results in a device-to-host (D2H) memory copy, as the tensor value must be materialized on the host for string representation**.
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.h#L821
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/_tensor.py#L584-L590
https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/_tensor_str.py#L336

### Versions

Collecting environment information...
PyTorch version: 2.8.0a0+git9450e38
Is debug build: True
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: version 3.31.0
Libc version: glibc-2.35

Python version: 3.10.15 | packaged by conda-forge | (main, Oct 16 2024, 01:24:24) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-182-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A10
Nvidia driver version: 535.161.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        3900.0000
CPU min MHz:                        1000.0000
BogoMIPS:                           4800.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: Split huge pages
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.15.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.14.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.0
[pip3] torch==2.8.0a0+git9450e38
[pip3] triton==3.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.8.0a0+git9450e38           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
[conda] triton                    3.3.0                    pypi_0    pypi

cc @jgong5 @wenzhe-nrv @sanchitintel @chauhang @penguinwu @zou3519 @bdhirsh

**Comment by EikanWang** [comment]
If a valid overload exists, the error message becomes unnecessary. Therefore, I'm considering deferring the Tensor representation (i.e., `__repr__`) until no valid overload can be found.

https://github.com/pytorch/pytorch/blob/7482eb217c621749dc11413ca1ae114690a09c55/torch/csrc/jit/python/pybind_utils.cpp#L797-L806

**Comment by albanD** [comment]
From a closer look at this, this is actually expected. Do you have any use case where you need to call the packet and not a specific overload?

**Comment by EikanWang** [comment]
I'm unsure whether `torch.ops.aten` is intended to be an official or user-facing way to develop PyTorch programs. If it is exposed as a public programming interface, I'm concerned that users may not be aware of the differences between `torch.ops.aten` and standard torch APIs, which could lead to unintended performance overhead or misuse.

Ex1:
https://github.com/pytorch/pytorch/blob/ffd49d538e8fcd274555df018ea2ba2ad8523df8/torch/masked/maskedtensor/_ops_refs.py#L147
```bash
>>> torch.ops.aten.where.overloads()
['self', 'ScalarOther', 'ScalarSelf', 'Scalar', 'default', 'self_out']
```

Ex2:
https://github.com/pytorch/pytorch/blob/ffd49d538e8fcd274555df018ea2ba2ad8523df8/torch/_inductor/fx_passes/misc_patterns.py#L56
```bash
>>> torch.ops.aten.index.overloads()
['Tensor', 'Tensor_out', 'Tensor_hacked_twin', 'str', 'list_int', 'list_float', 'list_bool', 'list_Tensor', 'list_str']
```

**Comment by zou3519** [comment]
torch.ops.aten is "public" in the sense that it is exposed to the world and you will commonly encounter it when working with torch.compile/torch.export. We expect that most end users will not use torch.ops.aten but will instead use the corresponding torch.* apis, because those are well documented.

**Comment by EikanWang** [comment]
> We expect that most end users will not use torch.ops.aten but will instead use the corresponding torch.* apis, because those are well documented.

It makes sense as torch.* apis are well documented. Since we always encourage users to use torch.* rather than torch.ops.aten.*, the exposure of this issue is low. We can close this issue and reopen it if users experience the issue in the future. @zou3519 , @albanD , are you okay with that?

**Comment by zou3519** [comment]
Sure
